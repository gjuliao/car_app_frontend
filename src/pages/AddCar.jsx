/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { addCar } from '../redux/car/carSlice';
import styles from '../styles/AddCar.module.css';
import noImage from '../assets/images/no_image_available.png';

const AddCar = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [formData, setFormData] = useState({});
  const [isValid, setIsValid] = useState(false);

  const validateForm = () => {
    let valid = true;
    if (formData.model?.length === 0) valid = false;
    if (formData.model === undefined) valid = false;
    if (formData.brand?.length === 0) valid = false;
    if (formData.brand === undefined) valid = false;
    if (formData.price?.length === 0) valid = false;
    if (formData.price === undefined) valid = false;
    if (formData.year === undefined) valid = false;
    if (formData.color?.length === 0) valid = false;
    if (formData.color === undefined) valid = false;
    if (formData.accidents?.length === 0) valid = false;
    if (formData.accidents === undefined) valid = false;
    if (formData.is_electric === null) valid = false;
    if (formData.is_electric === undefined) valid = false;
    if (formData.image?.length === 0) valid = false;
    if (formData.image === undefined) valid = false;
    if (formData.description?.length === 0) valid = false;
    if (formData.description === undefined) valid = false;
    setIsValid(valid);
  };

  const changeHandler = (e) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value,
    });
  };

  const sendForm = (e) => {
    e.preventDefault();
    dispatch(addCar(formData)).then((response) => {
      if (response.status !== 'error') navigate('/');
    });
  };

  useEffect(() => {
    validateForm();
  }, [formData]);

  return (
    <section className={styles.container}>
      <form className={styles.addForm} onSubmit={sendForm}>
        <input className={styles.formInput} type="text" name="model" placeholder="Model" onChange={changeHandler} />
        <input className={styles.formInput} type="text" name="brand" placeholder="Brand" onChange={changeHandler} />
        <input className={styles.formInput} type="text" name="price" placeholder="Price" onChange={changeHandler} />
        <input className={styles.formInput} type="date" name="year" placeholder="Year" onChange={changeHandler} />
        <input className={styles.formInput} type="text" name="color" placeholder="Color" onChange={changeHandler} />
        <input className={styles.formInput} type="text" name="accidents" placeholder="Accidents" onChange={changeHandler} />
        <select className={styles.formInput} type="text" name="is_electric" placeholder="Electric" onChange={changeHandler}>
          <option className={styles.nullOption} value={null}>
            Select: Electric or Gas
          </option>
          <option value>Electric</option>
          <option value={false}>Gas</option>
        </select>
        <input className={styles.formInput} type="text" name="image" placeholder="Image URL" onChange={changeHandler} />
        <img className={styles.photo} src={formData.image || noImage} alt="upload model" />
        <textarea className={styles.formText} name="description" placeholder="Add description here..." onChange={changeHandler} />
        <button className={styles.submit} type="submit" disabled={!isValid}>Create</button>
      </form>
    </section>
  );
};

export default AddCar;
