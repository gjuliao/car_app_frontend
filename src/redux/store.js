import { configureStore } from '@reduxjs/toolkit';
import reservationReducer from './reservation/reservationSlice';
import sessionReducer from './sessionSlice';
import carListReducer from './carlistSlice';
import carReducer from './carDetail/carDetailSlice';
import carItemsReducer, { fetchCars } from './car/carSlice';
import auth from './Auth/auth';

const store = configureStore({
  reducer: {
    auth,
    reservations: reservationReducer,
    session: sessionReducer,
    cars: carListReducer,
    car: carReducer,
    carlist: carItemsReducer,
  },
});
store.dispatch(fetchCars());
export default store;
