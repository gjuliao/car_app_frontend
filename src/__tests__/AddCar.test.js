import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import store from '../redux/store';
import AddCar from '../pages/AddCar';

describe('AddCar page tets', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <AddCar />
        </BrowserRouter>
      </Provider>,
    );
  });

  test('display create button', () => {
    expect(screen.getByText('Create')).toBeDefined();
  });

  test('Create button should be disabled', () => {
    expect(screen.getByText('Create')).toBeDisabled();
  });

  test('AddCar renders without crashing', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <AddCar />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
