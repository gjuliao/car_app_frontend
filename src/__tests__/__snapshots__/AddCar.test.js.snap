// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCar page tets AddCar renders without crashing 1`] = `
<section
  className="container"
>
  <form
    className="addForm"
    onSubmit={[Function]}
  >
    <input
      className="formInput"
      name="model"
      onChange={[Function]}
      placeholder="Model"
      type="text"
    />
    <input
      className="formInput"
      name="brand"
      onChange={[Function]}
      placeholder="Brand"
      type="text"
    />
    <input
      className="formInput"
      name="price"
      onChange={[Function]}
      placeholder="Price"
      type="text"
    />
    <input
      className="formInput"
      name="year"
      onChange={[Function]}
      placeholder="Year"
      type="date"
    />
    <input
      className="formInput"
      name="color"
      onChange={[Function]}
      placeholder="Color"
      type="text"
    />
    <input
      className="formInput"
      name="accidents"
      onChange={[Function]}
      placeholder="Accidents"
      type="text"
    />
    <select
      className="formInput"
      name="is_electric"
      onChange={[Function]}
      placeholder="Electric"
      type="text"
    >
      <option
        className="nullOption"
        value={null}
      >
        Select: Electric or Gas
      </option>
      <option
        value={true}
      >
        Electric
      </option>
      <option
        value={false}
      >
        Gas
      </option>
    </select>
    <input
      className="formInput"
      name="image"
      onChange={[Function]}
      placeholder="Image URL"
      type="text"
    />
    <img
      alt="upload model"
      className="photo"
      src=""
    />
    <textarea
      className="formText"
      name="description"
      onChange={[Function]}
      placeholder="Add description here..."
    />
    <button
      className="submit"
      disabled={true}
      type="submit"
    >
      Create
    </button>
  </form>
</section>
`;
